### ISSUE TITLE
Installing Gecko V0.0.8 on fresh HA install

### ISSUE THREAD

**Issue by simonk1969** [gazoodle/geckolib#27]
I've recently done a fresh installation of HA and this evening decided to try and set up the Gecko integration.

I'm installing through HACS, installation completes and then I reboot, when I then go to add integration I get a blank box popup waiting for input with a submit button (see picture), not sure what it is wanting but I tried the IP of Hot Tub, it then comes back with a box that show Hot Tub and a Submit button (see picture), press Submit and its sits there for ages with spinning icon and then after quit a while it comes back and says Created Configuration for Hot Tub (see picture), click Finish.  The intergration then shows Failed to set up (see picture). If I click on configure I see a options windows with six tick boxes but no writing (see picture).  If I check the logs there are lots of Gecko errors and warnings (see picture).

Any idea as to what is wrong or what I've done wrong?

Thanks
Simon

![Gecko 6](https://user-images.githubusercontent.com/24876982/152610790-7d4e6ba3-658e-4fcb-96a1-70d1c2d23de3.png)
![Gecko 5](https://user-images.githubusercontent.com/24876982/152610791-fa117871-0404-4540-ae18-e8894553d6a7.png)
![Gecko 4](https://user-images.githubusercontent.com/24876982/152610792-b6164533-8514-4a59-8664-7a84ffc88ca4.png)
![Gecko 3](https://user-images.githubusercontent.com/24876982/152610793-e515dfa1-f379-44b7-8c23-e0bb13845607.png)
![Gecko 2](https://user-images.githubusercontent.com/24876982/152610794-6de94706-6993-4c45-be97-6a9e4a41e4ef.png)
![Gecko 1](https://user-images.githubusercontent.com/24876982/152610795-23d0ac68-cb14-447b-a112-66e013d2a36a.png)

## System Health

version | core-2022.2.1
-- | --
installation_type | Home Assistant OS
dev | false
hassio | true
docker | true
user | root
virtualenv | false
python_version | 3.9.7
os_name | Linux
os_version | 5.10.91
arch | x86_64
timezone | Europe/London

<details><summary>Home Assistant Community Store</summary>

GitHub API | ok
-- | --
Github API Calls Remaining | 4846
Installed Version | 1.21.0
Stage | running
Available Repositories | 970
Downloaded Repositories | 4

</details>

<details><summary>Home Assistant Cloud</summary>

logged_in | true
-- | --
subscription_expiration | 27 February 2022, 00:00
relayer_connected | true
remote_enabled | true
remote_connected | true
alexa_enabled | true
google_enabled | true
remote_server | eu-west-2-0.ui.nabu.casa
can_reach_cert_server | ok
can_reach_cloud_auth | ok
can_reach_cloud | ok

</details>

<details><summary>Home Assistant Supervisor</summary>

host_os | Home Assistant OS 7.2
-- | --
update_channel | stable
supervisor_version | supervisor-2022.01.1
docker_version | 20.10.9
disk_total | 31.3 GB
disk_used | 6.6 GB
healthy | true
supported | true
board | ova
supervisor_api | ok
version_api | ok
installed_addons | SSH & Web Terminal (10.0.2), Home Assistant Google Drive Backup (0.105.2), Node-RED (10.4.0), File editor (5.3.3), ESPHome (2022.1.3)

</details>

<details><summary>Lovelace</summary>

dashboards | 1
-- | --
resources | 0
views | 1
mode | storage

</details>

**Comment by gazoodle** [comment]
Yep, my fault. For some reason v0.0.8 got a buggy version of en.json (no idea how it happened), but I've fixed that in v0.0.9. Try to do another update, you might need to force the integration in the HACS panel to locate the v0.0.9 release using the `Update Information" menu

![image](https://user-images.githubusercontent.com/47351872/152611446-4f5f75d2-8e06-4c9e-a1e1-63ec2e97284c.png)

**Comment by simonk1969** [comment]
That fixed it thankyou :-)

I never could get my Hot Tub working with this previously trying to get the spa xml file, etc.

Thanks for your great work, really appreciated :-)

**Comment by simonk1969** [comment]
![image](https://user-images.githubusercontent.com/24876982/152705792-a3e9ac79-f0dc-4c42-8852-ffde0e506079.png)
Since updating to V0.0.10 today I'm getting similar errors to yesterday :-(

**Comment by gazoodle** [comment]
Is the spa integration working as expected? Can you control the device from HA? I see the errors, I have some in my logs too, it’s because the integration sometimes times out, but I think I’ve seen a pattern and can investigate next week.

**Comment by simonk1969** [comment]
No, justs shows everything as UNAVAILABLE.

Work great since installing 0.0.9 two days ago until update to 0.0.10.

If you need any more info or need me to test something then just let me know :)

**Comment by gazoodle** [comment]
Have you restarted HA? Does it do the same each time?

**Comment by simonk1969** [comment]
I have restarted HA but still the same, all entities show unavailable.
I've also deleted integration and tol Gecko in HACS to redownload 0.0.9 but after restarting HA and adding integration the issue now remains the same.
Also deleted integration and Gecko install, reinstalled Gecko(tried both versions) added the integration but same result.

**Comment by gazoodle** [comment]
Hmmm. Sorry you're having such a mare with it. I did some changes earlier on today which I'm just about to commit which deals with some issues when there are retries during the class init after the spa has been configured. I originally thought that this might be the cause of your problem, but I'm not sure. Would you mind uploading the full log file rather than just the HA abridged version from above and I can see if it's similar to a problem I experienced earlier but was unable to reproduce (some exception about index out of range in an accessor)?

**Comment by gazoodle** [comment]
Also, can you try v0.0.11? Still interested in the log files to see what's behind the ERRORs.

**Comment by simonk1969** [comment]
I've just done a reboot and this was the results in the full ha log file:

`2022-02-07 19:45:17 WARNING (SyncWorker_1) [homeassistant.loader] We found a custom integration kia_uvo which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:17 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration nodered which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:17 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration hacs which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:17 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration gecko which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:25 WARNING (MainThread) [aioesphomeapi.reconnect_logic] Can't connect to ESPHome API for sonoff_4chpror2 @ 192.168.30.17: Error connecting to ('192.168.30.17', 6053): [Errno 113] Connect call failed ('192.168.30.17', 6053)
2022-02-07 19:45:30 WARNING (MainThread) [homeassistant.config_entries] Config entry 'WLED' for wled integration not ready yet: Invalid response from API: Error occurred while communicating with WLED device at 192.168.30.86; Retrying in background
2022-02-07 19:46:06 ERROR (Thread-5) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:46:06 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:46:18 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: gecko
2022-02-07 19:46:51 ERROR (Thread-8) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:46:51 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:47:18 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: gecko
2022-02-07 19:47:36 ERROR (Thread-11) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:47:36 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:48:18 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: gecko
2022-02-07 19:48:21 ERROR (Thread-14) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:48:21 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:48:21 ERROR (MainThread) [custom_components.gecko] Exception during entry setup
Traceback (most recent call last):
File "/config/custom_components/gecko/__init__.py", line 68, in async_setup_entry
raise Exception("Too many retries")
Exception: Too many retries`

**Comment by simonk1969** [comment]
I've just done a reboot and this was the results in the full ha log file:

2022-02-07 19:45:17 WARNING (SyncWorker_1) [homeassistant.loader] We found a custom integration kia_uvo which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:17 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration nodered which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:17 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration hacs which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:17 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration gecko which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-02-07 19:45:25 WARNING (MainThread) [aioesphomeapi.reconnect_logic] Can't connect to ESPHome API for sonoff_4chpror2 @ 192.168.30.17: Error connecting to ('192.168.30.17', 6053): [Errno 113] Connect call failed ('192.168.30.17', 6053)
2022-02-07 19:45:30 WARNING (MainThread) [homeassistant.config_entries] Config entry 'WLED' for wled integration not ready yet: Invalid response from API: Error occurred while communicating with WLED device at 192.168.30.86; Retrying in background
2022-02-07 19:46:06 ERROR (Thread-5) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:46:06 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:46:18 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: gecko
2022-02-07 19:46:51 ERROR (Thread-8) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:46:51 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:47:18 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: gecko
2022-02-07 19:47:36 ERROR (Thread-11) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:47:36 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:48:18 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: gecko
2022-02-07 19:48:21 ERROR (Thread-14) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.9/threading.py", line 910, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 310, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 340, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.9/site-packages/geckolib/spa.py", line 335, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2022-02-07 19:48:21 WARNING (MainThread) [custom_components.gecko] Facade went into error, lets retry
2022-02-07 19:48:21 ERROR (MainThread) [custom_components.gecko] Exception during entry setup
Traceback (most recent call last):
File "/config/custom_components/gecko/__init__.py", line 68, in async_setup_entry
raise Exception("Too many retries")
Exception: Too many retries

**Comment by simonk1969** [comment]
Unfortunately  v0.0.11 has issue for me.

I'm going to restore a backup from before I installed Gecko the other day and try 0.0.11 from there.

**Comment by simonk1969** [comment]
> 

Unfortunately didn't resolve it either, now not sure if I've got a problem with the unit or not!

I've rebooted both the hot tub and Gecko receiver, have a blue light on the receiver and can see it has a IP on my UDM Pro.

Any way to find more details on what is going on, what point it's failing?

**Comment by gazoodle** [comment]
Time for a full debug in the shell I think. 

Open a command prompt and 

```
pip install geckolib
python3
>>> from geckolib import GeckoShell
>>> GeckoShell.run(["logfile client.log", ‘loglevel DEBUG’])

  :
  :
```

Then upload the client.log file here. If your tub is not on the same subnet, you need to add `, ’discover IP-ADDRESS-OF-TUB’` to the shell run commands.

**Comment by simonk1969** [comment]
[client.log](https://github.com/gazoodle/geckolib/files/8022034/client.log)

client.log file attached.

Appreciate your time looking into :-)

**Comment by gazoodle** [comment]
Thanks. That log file is confusing!
The discover broadcast on your LAN is responded to correctly by your spa, but there are no responses coming back from any of the directed packets after that … the handlers basically run out of retries.  

What is confusing is that you appeared to have it working in v0.0.9 but not v0.0.10 but those versions had no change to geckolib which is where the protocol is failing!

I presume your iOS/Android in.touch2 app works fine, and if it does, that’s even more confusing. 

Unless, you’ve somehow got two gecko clients on the same machine - did you run that shell on your HA box?

if so, you probably need to remove the HA integration and restart HA to ensure it’s not still loaded.  

Otherwise 🤔… brain whirr …

**Comment by gazoodle** [comment]
Ok, wasn't that then. The logs are basically the same. So ...

The app works fine so it's unlikely anything to do with the in.touch2 module
The shell doesn't work, nor does HA, but they are on the same box so might be something to do with that or firewall or Unifi routing rules. 

Do you have any other wired or wireless devices that could potentially be used to run the shell to see if there is any difference there?

**Comment by simonk1969** [comment]
I've created a Ubuntu VM and ran the test again, log file attached
[client.log](https://github.com/gazoodle/geckolib/files/8023545/client.log)
.

**Comment by gazoodle** [comment]
Same result I'm afraid. 

If the VM is on the same box that we've previously been working on I'm not sure that will get around any networking issues with your host which is where I'm currently thinking we need to investigate ...

But if it's on a different box we need to understand why broadcast UDP from the box gets to the in.touch2 module and back to the host but point-to-point UDP packets aren't.

It's almost as if sending from the host to the in.touch2 module is being blocked somewhere but the broadcast (and the response back) is getting through fine.

I wonder, do you have multiple subnets on your network and if somehow the UDP broadcast is making it through and the packet from the in.touch2 module to your box, but the directed ones from the box to the in.touch2 module aren't

Can you PING the device from the VM and get a response back?

**Comment by simonk1969** [comment]
Yes the VM is on the same Proxmox box, which HA is also on.  I'll try a setup a Ubuntu physical box a bit later and see if I get the same results.

I'll try and do a bit more research about my UDM Pro with regards to UDP and specifically point to point UDP packets.

I do have multiple subnets VLANs set up all of the devices I'm trying this with are on my main subnet 192.168.1.x. 

I can ping the InTouch2 in get a response.

**Comment by gazoodle** [comment]
Something else quick to check, if you start the shell and try to get it to connect directly to your spa what happens.

```
>>> GeckoShell.run(["logfile client.log", ‘loglevel DEBUG’, 'discover 192.168.1.75'])
```

This won't do any broadcasting but will try to connect direct to the spa.

**Comment by simonk1969** [comment]
Looks to be the same.

 

Should have a physical Ubuntu box ready shortly to try.

**Comment by gazoodle** [comment]
That is very strange. Is the log file exactly the same? I wouldn't have expected the HELLO broadcast discover to have worked correctly.

The previous times you get
```
2022-02-08 10:33:44,355 geckolib.locator INFO Discovery process started
2022-02-08 10:33:44,356 geckolib.locator DEBUG Locator retry thread started
2022-02-08 10:33:44,406 geckolib.driver.udp_socket DEBUG Sending b'<HELLO>1</HELLO>' to ('<broadcast>', 10022)
2022-02-08 10:33:45,363 geckolib.driver.udp_socket DEBUG Sending b'<HELLO>1</HELLO>' to ('<broadcast>', 10022)
2022-02-08 10:33:45,375 geckolib.driver.udp_socket DEBUG Received b'<HELLO>SPAd8:80:39:a1:b0:44|Hot Tub</HELLO>' from ('192.168.1.75', 10022)
2022-02-08 10:33:46,382 geckolib.driver.udp_socket DEBUG Sending b'<HELLO>1</HELLO>' to ('<broadcast>', 10022)
2022-02-08 10:33:46,394 geckolib.driver.udp_socket DEBUG Received b'<HELLO>SPAd8:80:39:a1:b0:44|Hot Tub</HELLO>' from ('192.168.1.75', 10022)
2022-02-08 10:33:47,400 geckolib.driver.udp_socket DEBUG Sending b'<HELLO>1</HELLO>' to ('<broadcast>', 10022)
2022-02-08 10:33:47,412 geckolib.driver.udp_socket DEBUG Received b'<HELLO>SPAd8:80:39:a1:b0:44|Hot Tub</HELLO>' from ('192.168.1.75', 10022)
2022-02-08 10:33:48,368 geckolib.locator INFO Found 1 spas ... [Hot Tub(SPAd8:80:39:a1:b0:44)]
```

I was hoping that there would have been some difference. Sigh.

Well, lets see what your new box shows. Thanks for keeping at this with me.

**Comment by simonk1969** [comment]
[client.log](https://github.com/gazoodle/geckolib/files/8025189/client.log)
[clientwithIP.log](https://github.com/gazoodle/geckolib/files/8025196/clientwithIP.log)
Finally got a standalone box running.
Attached is client.log and also clientwithIP.log which is with the IP specific in command line.

**Comment by gazoodle** [comment]
Thanks @simonk1969. I'm really not sure what is happening here, but there is a difference we can investigate. The locator class enables socket_opt to support broadcast whereas the spa client doesn't. I have no idea if this will make a difference, but can you edit the file `spa.py`, find the function `start_connect` and make the following change:

Find the line `self.open()` and after it, add `self.enable_broadcast()`. Python is fussy with tabs and spaces, so make sure in indent is with spaces.

Should look something like this after ...

```python
def start_connect(self):
        """ Connect to the in.touch2 module """
        self._connection_started = time.monotonic()
        # Identify self to the intouch module
        self.open()
        # Add the line below vvv
        self.enable_broadcast()
        # Add the line above ^^^
        self.queue_send(
            GeckoHelloProtocolHandler.client(self.descriptor.client_identifier),
            self.descriptor.destination,
        )
   : 
   : 
(Truncated)
```
Then try starting the client again.

**Comment by simonk1969** [comment]
This may be a stupid question but where do I find spa.py I’m looking in config/custom_components/gecko and config/ but cannot find it? 

 

I’m using File Editor from within HA.

 

I have a put a post on Facebook Ubiquiti group for advice as to what could be blocking the UDP packets?

**Comment by simonk1969** [comment]
> Find the line self.open() and after it, add self.enable_broadcast(). Python is fussy with tabs and spaces, so make sure in indent is with spaces.

I tried this earlier but lookslike same results.

I also setup the linux box and intouch2 totally seperate using a old router and it appears to be the same, it broadcasts gets a response starts trying to comunicate but gets no responses and times out, so this rules out my network.

I've also reset the intouch2 receiver and repaired, powered off the hot tub for 5 minutes to reboot the transmitter and started it up again.

The android app is still working.

This is just so weird, would say its the receiver if it wasn't for the Android app still working!

**Comment by gazoodle** [comment]
?????

Ok, time to get busy :rolls sleeves up: diag script methinks. 

Also just to make sure, can you dig out the intouch2 module versions.  Should be in the app, related to your tub, I’m not sure precisely as I have iOS version. 

Should look like

intouch version EN 70 v14.0
intouch version CO 69 v11.0

**Comment by gazoodle** [comment]
Another thought occurs, maybe the in.touch module has blocked the app GUID for some reason … don’t see why, but you could also tweak that in driver/shell.py

Change the UUID, generate a new one at https://www.uuidgenerator.net/

`SHELL_UUID = "02ac6d28-42d0-41e3-ad22-274d0aa491da" `

**Comment by gazoodle** [comment]
Or we’re assuming the android app is communicating locally, but if it’s remote then that would invalidate our assumptions. 

If you’ve got PiHole or some other way to block a DNS lookup, block `intouch.geckoal.com` which will force your app to try locally, then we can confirm that the LAN protocol works. 

This is essentially what I had to do while sniffing the conversation early on when building this library.

**Comment by simonk1969** [comment]
> Another thought occurs, maybe the in.touch module has blocked the app UUID for some reason … don’t see why, but you could also tweak that in driver/shell.py
> 
> Change the UUID, generate a new one at https://www.uuidgenerator.net/
> 
> `SHELL_UUID = "02ac6d28-42d0-41e3-ad22-274d0aa491da" `

I'm not a linux expert, quite new to it so hopfully Im in the correct place but I cannot find shell.py
![image](https://user-images.githubusercontent.com/24876982/153182242-7bcd3293-5204-4937-906b-07db4fa8a27f.png)

**Comment by simonk1969** [comment]
> Or we’re assuming the android app is communicating locally, but if it’s remote then that would invalidate our assumptions.
> 
> If you’ve got PiHole or some other way to block a DNS lookup, block `intouch.geckoal.com` which will force your app to try locally, then we can confirm that the LAN protocol works.
> 
> This is essentially what I had to do while sniffing the conversation early on when building this library.
I've added intouch.geckoal.com to the block addresses in AdGaurd, disabled mobile data on phone and checked that phone is getting DNS from ADGaurd.  The phone app still connects to the hot tub.

**Comment by gazoodle** [comment]
Sorry, that path is utils/driver/shell.py

**Comment by gazoodle** [comment]
Ok, I've put a simple diag program to help us try to get to the bottom of this. It needs a slightly different install in that you need to get this repo down onto your system. For simplicity's sake, create a folder for source code in your home dir and then clone the repo there.

```
gazoodle@ubuntudev:/~$ mkdir Source
gazoodle@ubuntudev:/~$ cd Source
gazoodle@ubuntudev:/~/Source/$ git clone http://github.com/gazoodle/geckolib
Cloning into 'geckolib'...
warning: redirecting to https://github.com/gazoodle/geckolib/
remote: Enumerating objects: 1218, done.
remote: Counting objects: 100% (1218/1218), done.
remote: Compressing objects: 100% (533/533), done.
remote: Total 1218 (delta 823), reused 951 (delta 596), pack-reused 0
Receiving objects: 100% (1218/1218), 820.79 KiB | 508.00 KiB/s, done.
Resolving deltas: 100% (823/823), done.
gazoodle@ubuntudev:/~/Source/$ cd geckolib/tests
gazoodle@ubuntudev:/~/Source/geckolib/tests$ python3 diags.py
```

The diag tool does essentially what the shell does, but it's got a bit more debugging and we can fiddle around with it without breaking the library

You might need to install git too
```
$ sudo apt-get install git
```

**Comment by simonk1969** [comment]
Been a really busy day and looking like another one tomorrow, I'll try and continue with this tomorrow evening.

I really appreciate all the effort you are putting in with me to try and get it working for me :-)

**Comment by gazoodle** [comment]
You’re welcome. No hurry.

**Comment by simonk1969** [comment]
I'm wondering if there is a fault with the intouch2 so have ordered another should be here on monday.  We can prove then for sure if its the unit or not.  If it's not the unit then I can see about returning it or just sell it on.

**Comment by gazoodle** [comment]
This is very strange indeed. The diag.py program deliberately uses the same socket (which is still in broadcast mode). I assigned a new GUID so your in.touch2 module would never have heard of us before, plus it happily replied to the broadcast HELLO message but then failed to reply to a direct message ????

Also, earlier in this thread, you indicated that it was all working fine so I'm wondering what had changed since then? Maybe your in.touch2 module is playing up, but it's such an odd issue, why reply to the HELLO but not the PING!!!

So, lets think about this. There are the following devices in this setup

- PC
- Android phone
- Router
- in.touch2 LAN
- in.touch2 SPA
- A switch (I presume you don't have your in.touch2 module directly connected to the router)
- A wifi AP (Which when forced locally, the Android phone must be using)

We've tried changing the PC (by creating a new Ubuntu box instead of the HA box)
We've tried changing the router (by you putting in an old one instead of your UDM)
We've tried rebooting all the Gecko stuff (and you're going to try a different in.touch2, thanks that's very helpful)
We've not tried faffing with the switch (unless your simple setup was to plug the in.touch2 module direct into the router in which case nothing to see here).

The only other thing I can suggest is a wireshark session. This can be pretty hairy to setup but is the way that I did my original sniffing of packets to see what was going on. 

Also, it might be useful if you can describe in detail how all the pieces above are wired up or connected to see where we might put the wireshark probe if we decide to go down that route/

When I was doing my investigation, I had a python session running on my laptop in wifi mode so I could capture packets from both the laptop and the phone to compare - they were using the same AP.

If you use Unifi kit everywhere and have one of their APs, then this can be quite easy to capture the packets, I'll see if I can find the procedure which I put somewhere around here ... :-)

Meanwhile, as I've been typing this I thought of some other things we can try first, so I'll make some changes to diag.py and push them.

**Comment by gazoodle** [comment]
Ok, I've push some changes to the repo.

You can update your local copy by running the following command in the geckolib folder

```
git pull
```

Then you can try the diags again (I've increased the socket timeout to see if that helps) if that doesn't work then

```
python3 diag.py 192.168.0.100
```
Which will run the diags without a broadcast, but a direct connection to in.touch2 module

**Comment by simonk1969** [comment]
I've tried the above but still ends up with the same response.

I am doing all this testing on the Ubuntu box I made, connected to a ZyXell Router that I found stashed in the loft, factory reset.  intouch2 is also connect to this switch.  I can see in the router clients list that both have obtained a IP.  So it's just the three devices isolated from everything else, no internet.
 
Rather than wasting your time maybe we just wait until I get the other unit on monday and see if I get the same results with that?

**Comment by gazoodle** [comment]
Sounds like a plan, but before we do that, can you try this version? I was thinking about why it might not respond to us and one reason might be that it hasn't heard the registration hello message, so this diag.py shoves 5 of them at it just to make sure ... who knows.

So, same pattern, `git pull` followed by run the diags. If possible, can you put the output in reply just so I can see if there is any slight change in behaviour?

Thanks

**Comment by gazoodle** [comment]
Actually I'm not sure that will work either, I just changed the UUID in my diag.py and removed the HELLO handshake that I assumed was for making the module recognize a client, but the behaviour that you're getting seems like perhaps there is more to this than we think, maybe you've got a much newer firmware that has changed the protocol slightly?

Anyway, if this still happens on Monday, I'll need a wireshark capture when your Android app is doing it's handshake so I can see if that's what the problem is.

**Comment by simonk1969** [comment]
Received the new in.touch 2 kit this morning and swopped the house part and paired.  Ran the diag.py on the test box and instantly finds it and starts receiving lots of data :-)
Moved the house part back onto my network, Reinstalled the Gecko integration and all working in HA :-)
Replaced the spa part and repaired all still working okay :-)

Looks very much like there is something wrong with the house part of the Gecko kit and we've both been wasting our time trying to fix a hardware issue!

**Comment by gazoodle** [comment]
Well that’s good news, thanks for doing that. Glad you’ve got a working system. I wonder if you’d be happy to do a bit more investigation with the “not functioning” unit … because your Android app worked fine, and I’d like to understand why in case this isn’t down to faulty hardware but something I’m not doing quite right that your old unit objected to but the new one doesn’t?

**Comment by brewston** [comment]
I've been following this thread and it's been a great piece of troubleshooting. I'm slightly alarmed though that the expensive blue box could fail in the way that first one did and potentially stop the integration working :(

**Comment by gazoodle** [comment]
@simonk1969 If you can get the latest repo and try the diag tool again with the "old" in.touch2 module on your test lan that would be great. 

Over the weekend, my turgid memory recalled very early on the module was fussy around replying unless the client ID started IOS or AND (and I also think the UUID was in uppercase). Now, my unit no longer behaves like this, in fact it was happy to reply no matter what I made the client ID look like, but maybe the firmware on your unit didn't ...

Also, I looked back through this conversation and didn't find the intouch firmware versions I had asked for, can you do them again? On my IOS app, they are found by clicking on the cog icon when the app has connected to the spa, and then on Technical Info, in fact all that info would be useful.

Thanks

**Comment by simonk1969** [comment]
I took the screen shots but forgot to send them too you!

I've attached four screen shots, New Technical Info, New About and Old Technical Info, Old About.

![New Technical info](https://user-images.githubusercontent.com/24876982/153923051-95908f47-0658-435a-8857-5409c757ec1e.jpg)
![New About](https://user-images.githubusercontent.com/24876982/153923055-06e80d7a-720f-4755-a4e0-5cb711131187.jpg)
![Old Technical Info](https://user-images.githubusercontent.com/24876982/153923057-34b500fb-0b05-4b78-b0e1-c655e96aa741.jpg)
![Old About](https://user-images.githubusercontent.com/24876982/153923060-2f46663d-9726-4b8e-a832-13d840b43b3b.jpg)
!

**Comment by simonk1969** [comment]
[client.log](https://github.com/gazoodle/geckolib/files/8074132/client.log)
Client.log file attached using latest repo with the "old" in.touch2 module on test lan.

**Comment by gazoodle** [comment]
@simonk1969 did you do a `git pull` first? The log looks like it’s still got lowercase UUID in it …

**Comment by simonk1969** [comment]
> @simonk1969 did you do a `git pull` first? The log looks like it’s still got lowercase UUID in it …
> 
> Also, it’s the diag.py tool, not the shell

simon@UbuntuTest:~/Source/geckolib/tests$ python3 diag.py
2022-02-15 23:06:04,752> INFO Start comms diag with []
2022-02-15 23:06:04,753> INFO Discovery process started, create socket
2022-02-15 23:06:04,753> DEBUG UDP Socket created, open it
2022-02-15 23:06:04,754> DEBUG Socket opened, set broadcast
2022-02-15 23:06:04,754> DEBUG Broadcast enabled, add hello handler
2022-02-15 23:06:04,754> DEBUG Hello handler added, queue a broadcast
2022-02-15 23:06:04,805> DEBUG Sending b'<HELLO>1</HELLO>' to ('<broadcast>', 10022)
2022-02-15 23:06:04,817> DEBUG Received b'<HELLO>SPAd8:80:39:a1:b0:44|Hot Tub</HELLO>' from ('192.168.1.201', 10022)
2022-02-15 23:06:04,818> INFO _on_discovered: GeckoHelloProtocolHandler(send_bytes=b'<HELLO>1</HELLO>', age=4.077700214111246e-05, has_timedout=False, should_remove_handler=False, timeout=0s, retry_count=0) (was_broadcast=False, client_id=None, spa_id=b'SPAd8:80:39:a1:b0:44', spa_name=Hot Tub ), GeckoUdpSocket on <socket.socket fd=3, family=AddressFamily.AF_INET, type=SocketKind.SOCK_DGRAM, proto=17, laddr=('0.0.0.0', 34757)>
  receive_handlers=[GeckoHelloProtocolHandler(send_bytes=b'<HELLO>1</HELLO>', age=0.00036030900082550943, has_timedout=False, should_remove_handler=False, timeout=0s, retry_count=0) (was_broadcast=False, client_id=None, spa_id=b'SPAd8:80:39:a1:b0:44', spa_name=Hot Tub )],
  send_handlers=[]
  isopen: True isbusy: True, ('192.168.1.201', 10022)
2022-02-15 23:06:05,755> INFO Found spa : Hot Tub (b'SPAd8:80:39:a1:b0:44') at ('192.168.1.201', 10022), try to start connection
2022-02-15 23:06:05,756> DEBUG Connection started, install ping handler
2022-02-15 23:06:05,757> DEBUG Handler installed, send ping
2022-02-15 23:06:05,757> DEBUG Ping sent, wait for a minute
2022-02-15 23:06:05,784> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,836> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,887> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,939> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,990> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:06,042> DEBUG Sending b'<PACKT><SRCCN>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</SRCCN><DESCN>SPAd8:80:39:a1:b0:44</DESCN><DATAS>APING</DATAS></PACKT>' to ('192.168.1.201', 10022)
2022-02-15 23:07:05,796> ERROR Failed to get PING response
2022-02-15 23:07:05,796> INFO *** DIAG COMPLETE ***
simon@UbuntuTest:~/Source/geckolib/tests$

**Comment by simonk1969** [comment]
2022-02-15 23:06:04,752> INFO Start comms diag with []
2022-02-15 23:06:04,753> INFO Discovery process started, create socket
2022-02-15 23:06:04,753> DEBUG UDP Socket created, open it
2022-02-15 23:06:04,754> DEBUG Socket opened, set broadcast
2022-02-15 23:06:04,754> DEBUG Broadcast enabled, add hello handler
2022-02-15 23:06:04,754> DEBUG Hello handler added, queue a broadcast
2022-02-15 23:06:04,805> DEBUG Sending b'<HELLO>1</HELLO>' to ('<broadcast>', 10022)
2022-02-15 23:06:04,817> DEBUG Received b'<HELLO>SPAd8:80:39:a1:b0:44|Hot Tub</HELLO>' from ('192.168.1.201', 10022)
2022-02-15 23:06:04,818> INFO _on_discovered: GeckoHelloProtocolHandler(send_bytes=b'<HELLO>1</HELLO>', age=4.077700214111246e-05, has_timedout=False, should_remove_handler=False, timeout=0s, retry_count=0) (was_broadcast=False, client_id=None, spa_id=b'SPAd8:80:39:a1:b0:44', spa_name=Hot Tub ), GeckoUdpSocket on <socket.socket fd=3, family=AddressFamily.AF_INET, type=SocketKind.SOCK_DGRAM, proto=17, laddr=('0.0.0.0', 34757)>
  receive_handlers=[GeckoHelloProtocolHandler(send_bytes=b'<HELLO>1</HELLO>', age=0.00036030900082550943, has_timedout=False, should_remove_handler=False, timeout=0s, retry_count=0) (was_broadcast=False, client_id=None, spa_id=b'SPAd8:80:39:a1:b0:44', spa_name=Hot Tub )],
  send_handlers=[]
  isopen: True isbusy: True, ('192.168.1.201', 10022)
2022-02-15 23:06:05,755> INFO Found spa : Hot Tub (b'SPAd8:80:39:a1:b0:44') at ('192.168.1.201', 10022), try to start connection
2022-02-15 23:06:05,756> DEBUG Connection started, install ping handler
2022-02-15 23:06:05,757> DEBUG Handler installed, send ping
2022-02-15 23:06:05,757> DEBUG Ping sent, wait for a minute
2022-02-15 23:06:05,784> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,836> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,887> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,939> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:05,990> DEBUG Sending b'<HELLO>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</HELLO>' to ('192.168.1.201', 10022)
2022-02-15 23:06:06,042> DEBUG Sending b'<PACKT><SRCCN>IOSE31808F4-0CAC-462B-ADE6-55EE716DAD6E</SRCCN><DESCN>SPAd8:80:39:a1:b0:44</DESCN><DATAS>APING</DATAS></PACKT>' to ('192.168.1.201', 10022)
2022-02-15 23:07:05,796> ERROR Failed to get PING response
2022-02-15 23:07:05,796> INFO *** DIAG COMPLETE ***

**Comment by gazoodle** [comment]
Deep head scratch ... I have no idea what's going on. It's gotta be something about the directed message because all other communication works fine ... might have to do a Wireshark session if you're OK with that?

I kind of need to refresh my memory on how to drive that so we don't do too much back-and-forwarding ... let me see if I can put some thoughts together on this.

Also we need to work out how to wire things together so we can capture packets from your Android app as it's communicating with the device. 

This could be quite a bit of work, so totally understand if you would rather just get on with your life as you've got a working system now :-)

**Comment by simonk1969** [comment]
Does the in.touch2 need the spa part plugged in for this testing?
I've put the old in.touch2 on my normal network so I can try to connect phone app to both.
The app shows both devices listed but the old device shows 'Error The in.touch has lost connection with your spa' and the new one shows fine.

I've contacted the supplier and manufacturer about seeing if I can get a replacement for this faulty unit.

Where abouts in the world are you?

**Comment by gazoodle** [comment]
I think it will do the primary part of the conversation, but then replies with RFERR for when it can't contact the spa unit as this would be an expected condition, especially if the spa is powered off for maintenance or out-of-season.

I'm in south-east UK.

**Comment by simonk1969** [comment]
To be honest now that I've got a working system, I'm not worried about why the old unit didn't work, it wasn't 100% reliable on the app either and I'll just put it down to a faulty unit.

However if you want to investigate it further then I'm more than willing to work with you or I could send it to you for a little while before hopefully getting a replacement sorted that I can then sell on.

I don't mind either way as I very much appreciate the work that you have put in with me :-)

On a side note I'm in Hertfordshire, UK and was wondering what do you find the best approach for running Hot Tub in the UK?  Do you leave it at full temperature all the time, turn it down for certain hours?

Thanks
Simon

**Comment by gazoodle** [comment]
I'd be happy to take a closer look at the unit to see it it is intermittent, or something that affects the version of firmware on your device. Email me at gazoodle@hash.fyi and then I'll give you my address.

WRT running in the UK, the whole reason that I started this project was so that I could monitor energy use and potentially take advantage of split rate energy supply, heating it during the time when the energy was cheaper ... of course, most of them are now bust so it's not so much of an issue at the moment.

Early days, some of the folk I corresponded with were very interested for the same reasons, and I think they actually use the library for just that reason, although I got the impression that they lived in places where the temperature variation was much more varied than here in the UK.

**Comment by gazoodle** [comment]
Ok, thanks @simonk1969. I got the unit and, well here's the results.

- The unit behaves exactly as you said it did which is great, it means it's the unit and nothing else.

- I paired it with my spa unit and that process worked fine (although it never managed to get a solid blue light, it remained on green). I did a wireshark capture of all the packets to and from the unit and it resolutely refused to change it's behaviour, even with several reboots and re-pairings. I'm thinking now that coupled with your intermittent success with v0.0.8 and your Android app, the unit may well be faulty.

It's good that you have a new unit that works both with the integration and with the app, so I think we're both satisfied that you have a working system and while it's a shame not to have a better understanding of why the unit didn't work as expected, I'm of a mind to close this long thread and hope no one else gets the same behaviour as you.

For reference though, this happened with in.touch EN 88 v4.0 and CO 88 v3.0 in case anyone googles those numbers in future.
